Codemon Fetch

Enigmage opened the pull request:
Closes #17

#### Additions.

- `CodemonFetch` Module
  - Compatible with #12, #13 and #14 as the `make_structure` function in `CodemonFetch` module first checks if the necessary files and directories exist and then makes them if they don't. Once #12 #13 #14 are integrated into the project there would be no need to make any changes in this module as `make_structure`  function can then only as a safety check for the presence of Input and Output files.
  - `fetch_tests` function takes the contest name as argument and scrapes codeforces problem section to get the sample test cases and write them to input and output files (`A.in` and `A.op`  etc.) of respective problems.
  - if wrong contest name/number is provided, an error message is displayed.
- Dependencies
  - Added `requests` and `bs4` for scraping. 
  - Added `tqdm` for progress bar while fetching sample test cases.

#### Modifications
- `codemon.py`:
  - Added conditional to read fetch command. 
    ```py
        elif arg == "fetch":
        try:
          # if User provides contest number as arguement.
          fetch_tests(sys.argv[countArg])
        except IndexError:
          # Extract all numbers from current directory name.
          contest_number = ''.join(re.findall(r'\d+', os.getcwd().split('/')[-1]))
          fetch_tests(contest_number)
    ```
  - if user provides contest number as argument it will be simply passed to `fetch_tests` function. Otherwise there will be an
IndexError. In that case the user hasn't provided an argument. So, the contest number will be parsed from the current contest directory. Here, number of the contest is extracted using regex. Assumption is that the correct contest number should be present in the contest directory name. 
 
- `CodemonMeta` : 
  - A new way of extracting filenames was added to solve an inconsistency with scraping sample test cases. For example, contest number `1234` has problems `A, B1, B2, C, D, E, F`. So by the old way, sample test cases for problem B2 might get stored under the name of `C`. Or, in other cases where there are problems till `G` the sample test cases for the last problem will be missing.  
  - So to create a consistent interface for the user, file names will be extracted by scraping the contest page. Again, same assumption applies, the contest number should be present in the contest name provided by the user. If contest number is not there or if it is wrong, a generic `A, B, C, D, E, F` is used for filenames.

Enigmage committed:
codemon fetch

lucifer0987 committed:
Merge pull request #1 from ankingcodes/master

sync

lucifer0987 committed:
added codemon reg

Enigmage committed:
test

Enigmage committed:
fetch sample tests

Enigmage committed:
codemon fetch

Enigmage committed:
codemon fetch

Enigmage committed:
change output file extension

Enigmage committed:
error checks and help section updation

ankingcodes commented:
@Enigmage - I love this idea of parsing the questions from Codeforces and creating the files accordingly. I haven't reviewed your code yet but I'd be amazed if you could extend your functionality to create the directory structure described in #48, and create the files based on flag `-java`, `-py`, `-cpp` (default, if no flag then cpp files). In that case, I can award you the points of all three issues along with bonus points for coming up with this idea.

ankingcodes commented:
Check my [comment](https://github.com/ankingcodes/codemon/pull/30#discussion_r533976074) in the review for a better way to parse arguments from the shell. The `isCppFile` variable should be originally be `True` since if no flag is given, then we directly create Cpp files.,

ankingcodes commented:
Did some review and testing, you need to make the following changes:
- `codemon fetch` is serving two purposes, i.e., somewhat replaces `codemon init`, so I suggest adding a flag `-f` and `--fetch` along with `codemon init`, written as `codemon init --fetch 1455`, which would fetch all file names and testcases and create the recommended files. If only `codemon init` is used, create default directories (A,B, ..., G). 
- If user adds flag `-py` , `-java` or `-cpp` then create files with extension and template corresponding to them which will be declared in Codemon Meta. If no flag is given, proceed with C++ files. 
- Add verbose messages, such as `Creating file {somename}`, `Creating directories for contest`, `Fetching inputs and outputs` etc.

lucifer0987 committed:
updated codemon reg

Enigmage commented:
@ankingcodes  thanks for the review. 
> I'd be amazed if you could extend your functionality to create the directory structure described in #48 , and create the files based on flag -java, -py, -cpp (default, if no flag then cpp files)

Yes sure I can add that. I actually wanted to but since there were conflicting PRs #48 and #50 I thought i'll have to make a separate PR for extending this functionality once those changes are merged. But if you want I can extend that functionality in this PR itself. It should pretty much solve #12 #13 and #14 and the of course the original issue to this PR. Still, if you plan to merge #48 and #50 before this, just ping me because then I'll have to make my additions over the changes made by those PRs, to extend this functionality.

>Check my comment in the review for a better way to parse arguments from the shell. The isCppFile variable should be originally be True since if no flag is given, then we directly create Cpp files.,

Sure, will follow this.

> codemon fetch is serving two purposes, i.e., somewhat replaces codemon init, so I suggest adding a flag -f and --fetch along with codemon init, written as codemon init --fetch 1455, which would fetch all file names and testcases and create the recommended files. If only codemon init is used, create default directories (A,B, ..., G).

Okay so just a follow up question on this. If only `codemon init` is used then codemon will make the default directory structure `A,B,......,G` . If i'm not wrong, there would be no need to fetch sample test cases in this case since the directory structure won't coincide with any contest on CodeForces, so in this case do you want me to initialize input and output (`.in` and `.op`) files?

> If user adds flag -py , -java or -cpp then create files with extension and template corresponding to them which will be declared in Codemon Meta. If no flag is given, proceed with C++ files.
Add verbose messages, such as Creating file {somename}, Creating directories for contest, Fetching inputs and outputs etc.

Okay will take care of this.

alphaX86 commented:
@Enigmage does it resolve issues without merging other templates or do I need to change any of them in my PR before merging yours? I'm asking this doubt to avoid conflict in mine (#50 ) and in other branches. Thanks

Enigmage commented:
@alphaX86 if @ankingcodes merges your PR first, you won't have have to change anything, I will make the necessary additions over your changes. This is why I have asked which ones he is planning to merge first. Let's see what @ankingcodes has to say.

ankingcodes commented:
>  I have a very good generic solution in mind that will be able to initialize directory structure for all three supported languages with very little code addition, It should pretty much solve #12 #13 and #14 and of course the original issue to this PR

@Enigmage - I'm a fan of less code doing more things. Your PR is already doing that quite creatively. I don't think it will be a problem to extend the functionality of this PR to what #50 #48 are trying to do. 

You need not worry about the merging of other PRs. If you have a better idea, you should always present it in your PR. It's the responsibility of the maintainer to access other PRs and merge the best one. For example, other PRs are right now doing the same thing you're trying to do, but you have a solution which is better and requires less code than those PRs (I had the same thing in my mind too, was expecting someone to get it done). So go ahead and implement those. 

Just to clarify things, I'd reiterate the solution again. 
- `codemon init` (without `-f` or `--fetch`) : create directory structure normally (A,B,C...,G), default in C++ else Python or Java corresponding to the flags `-py` or `-java` given by user as arguments.
- `codemon init --fetch` or `codemon init -f`: create directory structure by scraping data from codeforces. Remember to create test case files. 
- (optional) `codemon init -n `: create a single file with given name, if possible parse the problem and contest number (eg: `1455B` should have problem name B and contest number 1455) and if user gives flag `--fetch` or `-f` then fetch the input output and create the necessary files. 
- (optional) Create an issue for scraping inputs outputs from codechef website, google codejam website, kickstart website, AtCoder website etc.

Enigmage commented:
Okay cool. I will then add the required changes and ping you for review.

lucifer0987 committed:
added changes as requested

Enigmage commented:
Hey @ankingcodes can you confirm what exactly the directory structure should look like after running init. I referred to your comments on various issues and PRs but its still not completely confirmed. This is what I'm getting at:
```
contest_name/
  A /
    A.cpp
    A.in
    A.op
    test_case.cpp
    test_case
  B /
......
```
If there is anything missing or wrong do point out. I'm done with parsing part and almost done with `init` just needed a confirmation of the final picture you had in mind. Thanks.

ankingcodes commented:
@Enigmage - I had discussed about having testcase files inside each directory, which would be cleaned up using `codemon clean`, but on second thoughts, I think that would be another effort for the user and the concept behind codemon is to keep user effort as minimum as possible, so as a result, I believe we should get rid of it from inside the problem directory and place it on the contest directory. That way even if a user doesn't run `codemon clean`, the amount of useless testcase files gets reduced.

Enigmage commented:
@ankingcodes thanks, got it.

lucifer0987 committed:
added asteriks to password input

lucifer0987 committed:
updated codemon reg

ankushbhardwxj committed:
Merge pull request #54 from lucifer0987/codemon-reg

Added command "codemon reg"
[ reviewed by @ankingcodes ]

ankingcodes commented:
@Enigmage - resolve the conflicts 👀

Enigmage committed:
efficiently parse command-line options

Enigmage committed:
remove test file

Enigmage committed:
merge-upstream

Enigmage commented:
@ankingcodes Done.

ankushbhardwxj commented on setup.py:
```diff
@@ -9,6 +9,9 @@
     'watchdog',
     'clint',
     'stdiomask'
```
getting an error in installing, you missed a `,` here

ankingcodes commented:
@Enigmage - package installation fails

ankingcodes commented:
@Enigmage - I think you made some mistake while resolving conflicts. I dont see any of your latest commits of the changes I had requested.
Look into it.

Enigmage commented:
@ankingcodes I just made the commits that resolved the conflicts. I will be making the changes you requested soon.
>getting an error in installing, you missed a , here

yes you are right I'll resolve it wait.

Enigmage committed:
reslove mistake in setup.py

Enigmage commented:
@ankingcodes resolved the issue, installation should be working now. I will be making the requested commits soon.

Enigmage committed:
initializations

Enigmage committed:
fetch

Enigmage committed:
resolve-conflicts

ankingcodes commented:
@Enigmage - Ping me for review once you complete the features mentioned in https://github.com/ankingcodes/codemon/pull/55#issuecomment-739752918

Enigmage committed:
requested changes added

Enigmage commented:
@ankingcodes  So the changes are done. You can review now.

closes #12 #13 #14 and #17

#### What I didn't add:

- Default templates for java and python. So that others who are interested can add them. I'll open issues with detailed instructions on how to add templates for the two.
- Listen behaviour. Since the directory structure is new and two new languages are added, a lot of work will be needed in Listening behaviour of codemon, since that's a separate issue, I haven't made a lot of changes there. But I have hacked together a way, for now, such that listen works for cpp files, takes input from `test_case` and compiles. We'll have to open issues for adding listening behaviour for Python and Java Files and also optimizing it for Cpp.
- I haven't added the third optional task you gave me about adding the fetch functionality to single files. With your approval, I would like to open a separate issue for it and work on it separately. 

#### What I added

- `Parsing`: Borrowed your Boolean approach. So there are three parts to your approach: first initializing the booleans, second setting their values based on command line arguments provided and third performing operations based on the booleans. So I wrapped the first two parts of your approach into a class `Parser`, booleans being the data members and and member function `parse` that takes `sys.argv` as argument and sets the values of those booleans. I added this class in a separate module so that the code is readable, modular and the `main` function only contains the third part of your approach. Also makes it easy for developers in the future to add new commands.
  - I have separated flags from non flag arguments in the `Parser` class so that flags don't have to be fixed at a certain position. 

- `Templates`: Added a Template class with four methods, three for default templates and one for checking `~/.codemon/` folder for custom templates. During init if custom templates are not there default are added. Example snippet:

```py
    if init_flags["is_py"]:
        ext, use_template = "py", templates.get_custom_template("py") or templates.default_py()

```
`Init`: Init functionality as you asked based on flags for all three filetypes. Didn't need to add any new function for this, just had to make the present `init` function accept a third argument `init_flags` which is a dictionary of booleans containing different options set during the parse phase. Based on that `init` is done for all three filetypes. Init for single files also added based on flags, same for practice files.

- ` Fetch`: Rest about fetch is same.  Added -f , --fetch flag/option to fetch and init at the same time.

Enigmage committed:
error checks and code refactoring

Enigmage commented:
@ankingcodes  if you're reviewing this, I have added a few more changes, error checks basically.

- Earlier the user could get away with typing commands like `codemon reg this xx` or `codemon listen abc` or ` codemon -py -java init` etc and there would be no errors. I have added error checks that will show help section if such faulty commands are entered.
  - If user enters two language flags at the same time, it will cause an error and show help section
  - if user enters extra arguments in commands like `listen` , `reg` that take only one argument, it will cause an error and show help section.

- In case the user did not have internet connection `fetch` would error out and contest directory won't be created. Now, I have added exception handling which in  of such a situation will create a default directory structure, so the user won't have to depend on internet connection for initialization of contest directory.

ankingcodes commented:
@Enigmage - Reviewed your PR. Excellent work ! Everything seems to be on point. Many thanks for refactoring code and going OOP style. There are a few tweaks I'd like to be made but I'm sure I can do that myself or maybe make separate issues for them. Thanks for contributing.
